File: camunda-embedded-test/src/main/java/io/github/thake/camundaembedded/EmbeddedCamundaTestExtension.java
package io.github.thake.camundaembedded;

import io.camunda.process.test.api.CamundaProcessTest;
import io.camunda.process.test.api.CamundaProcessTestExtension;
import io.camunda.process.test.api.CamundaProcessTestRuntimeMode;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.AfterEachCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.BeforeEachCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.platform.commons.support.AnnotationSupport;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Proxy;
import java.net.URI;
import java.util.Optional;

public class EmbeddedCamundaTestExtension implements
        BeforeAllCallback,
        AfterAllCallback,
        BeforeEachCallback,
        AfterEachCallback {

    public static final int DEFAULT_PORT = 0;

    private static ManagedCamundaProcess server;
    private static int refCount = 0;

    private final int grpcPort;
    private final int restPort;
    private final int monitoringPort;

    private CamundaProcessTestExtension cpt;

    @CamundaProcessTest
    private static final class CamundaAnnotatedDummy {
    }

    public EmbeddedCamundaTestExtension() {
        this(DEFAULT_PORT, DEFAULT_PORT, DEFAULT_PORT);
    }

    public EmbeddedCamundaTestExtension(int grpcPort, int restPort, int monitoringPort) {
        this.grpcPort = grpcPort;
        this.restPort = restPort;
        this.monitoringPort = monitoringPort;
    }

    public static synchronized ManagedCamundaProcess startServer(int grpcPort, int restPort, int monitoringPort) {
        if (server == null) {
            server = new ManagedCamundaProcess(grpcPort, restPort, monitoringPort).start();
        }
        refCount++;
        return server;
    }

    public static synchronized void stopServer() {
        refCount--;
        if (refCount <= 0 && server != null) {
            server.close();
            server = null;
            refCount = 0;
        }
    }

    @Override
    public void beforeAll(ExtensionContext context) throws Exception {
        EmbeddedCamundaTest annotation = AnnotationSupport.findAnnotation(context.getElement(), EmbeddedCamundaTest.class)
                .orElse(null);

        int targetGrpcPort = (annotation != null && annotation.grpcPort() != 0) ? annotation.grpcPort() : this.grpcPort;
        int targetRestPort = (annotation != null && annotation.restPort() != 0) ? annotation.restPort() : this.restPort;
        int targetMonitoringPort = (annotation != null && annotation.monitoringPort() != 0) ? annotation.monitoringPort() : this.monitoringPort;

        ManagedCamundaProcess runningServer = startServer(targetGrpcPort, targetRestPort, targetMonitoringPort);

        CamundaProcessTestExtension extension = new CamundaProcessTestExtension()
                .withRuntimeMode(CamundaProcessTestRuntimeMode.REMOTE)
                .withCamundaClientBuilderOverrides(builder -> {
                    builder.grpcAddress(URI.create("http://localhost:" + runningServer.getActualGrpcPort()))
                            .restAddress(URI.create("http://localhost:" + runningServer.getActualRestPort()));
                })
                .withRemoteCamundaMonitoringApiAddress(URI.create("http://localhost:" + runningServer.getActualMonitoringPort()));
        this.cpt = extension;

        ExtensionContext wrappedContext = (ExtensionContext) Proxy.newProxyInstance(
                ExtensionContext.class.getClassLoader(),
                new Class<?>[]{ExtensionContext.class},
                (proxy, method, args) -> {
                    if ("getTestClass".equals(method.getName()) && (args == null || args.length == 0)) {
                        return Optional.of(CamundaAnnotatedDummy.class);
                    }
                    if ("getRequiredTestClass".equals(method.getName()) && (args == null || args.length == 0)) {
                        return context.getRequiredTestClass();
                    }
                    try {
                        return method.invoke(context, args);
                    } catch (InvocationTargetException e) {
                        throw e.getCause();
                    }
                }
        );
        extension.beforeAll(wrappedContext);
    }

    @Override
    public void beforeEach(ExtensionContext context) throws Exception {
        if (cpt != null) {
            cpt.beforeEach(context);
        }
    }

    @Override
    public void afterEach(ExtensionContext context) throws Exception {
        if (cpt != null) {
            cpt.afterEach(context);
        }
    }

    @Override
    public void afterAll(ExtensionContext context) throws Exception {
        try {
            if (cpt != null) {
                cpt.afterAll(context);
            }
        } finally {
            cpt = null;
            stopServer();
        }
    }

    public int getGrpcPort() {
        return grpcPort;
    }

    public int getRestPort() {
        return restPort;
    }

    public int getMonitoringPort() {
        return monitoringPort;
    }
}

File: camunda-embedded-test/src/main/kotlin/io/github/thake/camundaembedded/EmbeddedCamundaTestExtension.kt
package io.github.thake.camundaembedded

import io.camunda.process.test.api.CamundaProcessTest
import io.camunda.process.test.api.CamundaProcessTestExtension
import io.camunda.process.test.api.CamundaProcessTestRuntimeMode
import org.junit.jupiter.api.extension.*
import org.junit.platform.commons.support.AnnotationSupport
import java.net.URI
import java.util.Optional

@CamundaProcessTest
private class CamundaAnnotatedDummy

class EmbeddedCamundaTestExtension(
    val grpcPort: Int = DEFAULT_PORT,
    val restPort: Int = DEFAULT_PORT,
    val monitoringPort: Int = DEFAULT_PORT
) : BeforeAllCallback,
    AfterAllCallback,
    BeforeEachCallback,
    AfterEachCallback {

    companion object {
        const val DEFAULT_PORT = 0

        private var server: ManagedCamundaProcess? = null
        private var refCount = 0

        @Synchronized
        fun startServer(
            grpcPort: Int = DEFAULT_PORT,
            restPort: Int = DEFAULT_PORT,
            monitoringPort: Int = DEFAULT_PORT
        ): ManagedCamundaProcess {
            if (server == null) {
                server = ManagedCamundaProcess(
                    grpcPort = grpcPort,
                    restPort = restPort,
                    monitoringPort = monitoringPort
                ).start()
            }
            refCount++
            return server!!
        }

        @Synchronized
        fun stopServer() {
            refCount--
            if (refCount <= 0 && server != null) {
                server?.close()
                server = null
                refCount = 0
            }
        }
    }

    private var cpt: CamundaProcessTestExtension? = null

    override fun beforeAll(context: ExtensionContext) {
        val annotation = AnnotationSupport.findAnnotation(context.element, EmbeddedCamundaTest::class.java).orElse(null)

        val targetGrpcPort = annotation?.grpcPort ?: grpcPort
        val targetRestPort = annotation?.restPort ?: restPort
        val targetMonitoringPort = annotation?.monitoringPort ?: monitoringPort

        val runningServer = startServer(targetGrpcPort, targetRestPort, targetMonitoringPort)

        val extension = CamundaProcessTestExtension()
            .withRuntimeMode(CamundaProcessTestRuntimeMode.REMOTE)
            .withCamundaClientBuilderOverrides { builder ->
                builder.grpcAddress(URI.create("http://localhost:${runningServer.actualGrpcPort}"))
                    .restAddress(URI.create("http://localhost:${runningServer.actualRestPort}"))
            }
            .withRemoteCamundaMonitoringApiAddress(URI.create("http://localhost:${runningServer.actualMonitoringPort}"))
        cpt = extension

        val wrappedContext = object : ExtensionContext by context {
            override fun getTestClass(): Optional<Class<*>> {
                return Optional.of(CamundaAnnotatedDummy::class.java)
            }
            override fun getRequiredTestClass(): Class<*> {
                return context.requiredTestClass
            }
        }
        extension.beforeAll(wrappedContext)
    }

    override fun beforeEach(context: ExtensionContext) {
        cpt?.beforeEach(context)
    }

    override fun afterEach(context: ExtensionContext) {
        cpt?.afterEach(context)
    }

    override fun afterAll(context: ExtensionContext) {
        try {
            cpt?.afterAll(context)
        } finally {
            cpt = null
            stopServer()
        }
    }
}
